### ISSUE TITLE
How should one deal with a NIL being returned after EXEC?

### ISSUE THREAD

**Issue by jhnwkmn** [redis/hiredis#922]
Hi,

I'm using hiredis 0.14.0 and when putting data to Redis I basically do it like:

  redisAppendCommand(..., "MULTI");
  redisAppendCommand(..., "SADD %b %b", ...);
  ...
  redisAppendCommand(..., "SET %b %b", ...);
  redisAppendCommand(..., "EXEC", ...);

Then I call redisGetReply as many times as needed. I always get an "OK" status
reply for the MULTI, followed by as many "QUEUED" status replies as I appended "SADD"s
and "SET"s and then finally an array reply for the EXEC, that then contains the actual
replies of the commands.

However, this occasionally fails for another user. Based on the logs of the program, it
appears that the first reply after "EXEC" is NIL. And after that it seems that the commands
sent by the program and the replies received are out of sync. Unfortunately I have not been
able to reproduce the behaviour, so I have no more data to work on that what the logging of
the program reveals.

A few questions.

- If the first reply read after "EXEC" is NIL, will there be additional replies that need
  to be read, in order to empty the reply buffer? If so, how many? Is the NIL the reply of
  "MULTI" or an extra indication that there was a failure?
- Is there a way to blindly discard all unread replies, if one simply chooses to give up
  at first failure?
- Is there a way to find out how many pending replies there are? In blocking mode, if
  there is no unread reply, calling redisGetReply(...) blocks the caller.

Thank you,
  Johan Wikman

**Comment by michael-grunder** [comment]
> If the first reply read after "EXEC" is NIL, will there be additional replies that need
to be read, in order to empty the reply buffer? If so, how many? Is the NIL the reply of
"MULTI" or an extra indication that there was a failure?

Unfortunately, this can occur for a variety of reasons (out of memory, timeout, etc).  The safest course of action would be to check the context for an error (`c->err`, `c->errstr`) and then reconnect.

> Is there a way to blindly discard all unread replies, if one simply chooses to give up
at first failure?

Before Redis 6.2 the only clean way to do this is to reconnect.  For 6.2 and newer you can use the [RESET](https://redis.io/commands/reset) command which was introduced for this very problem.  It's a common pain point across basically all of the client libraries.

> Is there a way to find out how many pending replies there are? In blocking mode, if
there is no unread reply, calling redisGetReply(...) blocks the caller

No, not that I know of.  I suppose it would be possible to set the socket to non-blocking mode, and then call `redisGetReply` until it returns NULL or go lower level and `epoll` the underlying fd manually.  This would make me nervous though since Redis could be handling an expensive command when you make your checks, only to send more replies later.

TL;DR:  Before Redis 6.2 reconnecting is the safe option and >= 6.2 you can use the RESET command.

**Comment by jhnwkmn** [comment]
@michael-grunder, thank you, this was most informative.

**Comment by michael-grunder** [comment]
No worries, I'm glad I could help.

I'm going to close the issue but feel free to post here if you have other questions!
